submissions/AsyaYeromina/Tiny-JS-world-OOP/index.js

```diff
@@ -0,0 +1,167 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+  Complete the below for code reviewers' convenience:
+
+  Code repository: https://github.com/AsyaYeromina/a-tiny-JS-world
+  Web app: https://asyayeromina.github.io/a-tiny-JS-world/.
+  */
+
+// ======== OBJECTS DEFINITIONS ========
+// Define your objects here
+
+class Inhabitant {
+  constructor(inhabitantProperties) {
+    this.species = inhabitantProperties.species;
+    this.name = inhabitantProperties.name;
+    this.gender = inhabitantProperties.gender;
+    this.legs = inhabitantProperties.legs;
+    this.speech = inhabitantProperties.speech;
+    this.friends = inhabitantProperties.friends;
+  }
+}
+
+class Animal extends Inhabitant {
+  constructor(inhabitantProperties) {
+    super({ ...inhabitantProperties, legs: 4});
+    this.tail = inhabitantProperties.tail;
+  }
+
+  toString() {
+    return [
+      this.species,
+      this.name,
+      this.gender,
+      this.legs,
+      this.tail,
+      this.speech,
+      this.friends,
+    ]
+    .map((propertyValue) =>
+    propertyValue == this.tail && propertyValue === true
+      ? (propertyValue = "has tail")
+      : propertyValue == this.tail && propertyValue === false
+      ? (propertyValue = "no tail")
+      : propertyValue
+    )
+    .map((propertyValue) =>
+      Array.isArray(propertyValue) ? propertyValue.join(", ") : propertyValue
+    )
+    .join("; ");
+    }
+}
+
+class Canis extends Animal {
+  constructor(inhabitantProperties) {
+    super({ ...inhabitantProperties, species: "Canis", speech: "Woof-woof!", tail: true });
+  }
+}
+
+class FelisCatus extends Animal {
+  static catSpeech = "Meow";
+
+  constructor(inhabitantProperties) {
+    super({ ...inhabitantProperties, species: "Felis catus", speech: FelisCatus.catSpeech, tail: true });
+  }
+}
+
+class Sapiens extends Inhabitant {
+  constructor(inhabitantProperties) {
+    super({ ...inhabitantProperties, legs: 2});
+    this.hands = 2;
+  }
+
+  toString() {
+    return [
+      this.species,
+      this.name,
+      this.gender,
+      this.legs,
+      this.hands,
+      this.speech,
+      this.friends,
+    ]
+      .map((propertyValue) =>
+        Array.isArray(propertyValue) ? propertyValue.join(", ") : propertyValue
+      )
+      .join("; ");
```

Comment:
The previous similar example sourced too much information. So I will use this one to emphasize yet another issue.

Imagine, we add winged birds, horned cattle, finned fish, nautic mammals etc, all introducing their unique features.

Current approach will appear suboptimal pretty soon. Encapsulation and methods overloading (with calls to parents' methods) are to the rescue.

Encapsulation: let each class handle props it **owns** directly.
This means that `Inhabitant` class should implement its own version of `toString` to build representation from the props it owns directly.
Same is applicable to every child class that introduces its own unique property. How a child class can benefit from parent's `toString` implementation? `super.methodName()` calls an inherited version of a method.

E.g.
```
toString() {
  return
    [ super.toString(),
      this.uniqueProperty,
    ].join(delimiterConst);
}
```

Perhaps, in `Animal` class `this.uniqueProperty` will work even better as `this.uniqueProperty ? "has tail" : "no tail"`

Well, yes. Tails and hands will come after everything else. A trade-off. Or consider reordering properties in the list so they combined still create a sense of sanity.

Explicit labels in presentation string might also work well.

Consider the example below:
```
["propA", "propB"].map(propName => `${propName}: ${this[propName]}`).join(delimiter);
```